node/src/actors/chain_manager/handlers.rs

```diff
@@ -306,118 +302,234 @@ impl Handler<AddBlocks> for ChainManager {
-                                    .and_then(move |_, _act, _ctx| actix::fut::ok(()))
-                                    .wait(ctx);
-                            }
+                if msg.blocks.is_empty() {
+                    log::debug!("Received an empty AddBlocks message");
+                    self.sm_state = StateMachine::WaitingConsensus;
+                    self.initialize_from_storage(ctx);
+                    log::info!("Restored chain state from storage");
 
-                            if let Err(e) = self.process_requested_block(ctx, block.clone()) {
-                                log::error!("Error processing block: {}", e);
-                                self.initialize_from_storage(ctx);
-                                log::info!("Restored chain state from storage");
-                                batch_succeeded = false;
-                                break;
-                            }
+                    return;
+                }
 
-                            let beacon = self.get_chain_beacon();
-                            show_sync_progress(
-                                beacon,
-                                target_beacon,
-                                self.epoch_constants.unwrap(),
-                            );
+                let sync_target = self.sync_target.clone().unwrap();
+                let superblock_period = u32::from(consensus_constants.superblock_period);
+
+                // TODO: Review this hack
+                // Hack: remove first block
+                // For some reason the other node sends one extra block, and we cannot consolidate
+                // it because it is our top block.
+                msg.blocks.remove(0);
+
+                // Split received blocks into batches according to 3 different cases:
+                // 1. TargetNotReached: superblock target not reachable with the received block batch.
+                // 2. SyncWithoutCandidate: target superblock needs to be consolidated, but no
+                // candidate superblock should be built.
+                // 3. SyncWithCandidate: target superblock needs to be consolidated, and a candidate
+                // superblock should be built.
+                let block_batches = split_blocks_batch_at_target(
+                    |b| b.block_header.beacon.checkpoint,
+                    msg.blocks,
+                    self.current_epoch.unwrap(),
+                    &sync_target,
+                    superblock_period,
+                );
 
-                            if beacon == target_beacon {
-                                break;
-                            }
+                use BlockBatches::*;
+                match block_batches {
+                    // TargetNotReached:
+                    // 1. process blocks
+                    // 2. requests block batch -> revert to WaitingConsensus
+                    Ok(TargetNotReached(blocks1)) => {
```

Comment:
lets just change the naming of blocks 1 blocks 2 and blocks 3 maybe? in case 1 this can be something like non_consolidable_blocks